extensions/ql-vscode/src/model-editor/shared/validation.ts

```diff
@@ -117,24 +117,40 @@ export function validateModeledMethods(
     }
   }
 
-  const neutralModeledMethod = consideredModeledMethods.find(
+  const neutralModeledMethods = consideredModeledMethods.filter(
     (modeledMethod) => modeledMethod.type === "neutral",
   );
-  const hasNonNeutralModeledMethod = consideredModeledMethods.some(
-    (modeledMethod) => modeledMethod.type !== "neutral",
-  );
 
-  // If there is a neutral model and any other model, that is an error
-  if (neutralModeledMethod && hasNonNeutralModeledMethod) {
-    // Another validation will validate that only one neutral method is present, so we only need
-    // to return an error for the first one
+  const neutralModeledMethodsByKind = new Map<string, ModeledMethod[]>();
+  for (const neutralModeledMethod of neutralModeledMethods) {
+    if (!neutralModeledMethodsByKind.has(neutralModeledMethod.kind)) {
+      neutralModeledMethodsByKind.set(neutralModeledMethod.kind, []);
+    }
+
+    neutralModeledMethodsByKind
+      .get(neutralModeledMethod.kind)
+      ?.push(neutralModeledMethod);
+  }
+
+  for (const [
+    neutralModeledMethodKind,
+    neutralModeledMethods,
+  ] of neutralModeledMethodsByKind) {
+    const conflictingMethods = consideredModeledMethods.filter(
+      (method) => neutralModeledMethodKind === method.type,
```

Comment:
Comparing `kind` and `type` here doesn't look right. We're also not filtering out the neutral method itself so the `conflictingMethods.length < 1` check below would likely always detect the neutral method.

Should this be
```suggestion
      (method) => method.type !== "neutral" && neutralModeledMethodKind === method.kind,
```

Reply:
I think this is correct. We're trying to match up the neutral's kind with the non-neutral's type. For example, we want to find these two conflicting classifications:

```json5
// Neutral classification
{
  "type": "neutral",
  "kind": "sink",
  // ...
}

// Sink classification
{
  "type": "sink",
  "kind": "code-injection",
  // ...
}
```

Therefore, the condition for a single conflict is something like `neutral.type === 'neutral' && nonNeutral.kind === neutral.type`.

We're explicitly not looking at the non-neutral method's kind since that's `code-injection`, `remote`, or `taint`, etc.

Reply:
I see. Thanks for explaining. I hadn't really understood / had forgotten how this worked. I now agree that this makes sense and is the right thing to do.